### ISSUE TITLE
您好，依据DCNN论文词向量没一维都会有CONV作用，可以请教一下CONV1D是如何作用在词向量的每一维上的吗？

### ISSUE THREAD

**Issue by 473951841** [yongzhuo/Keras-TextClassification#36]

**Comment by yongzhuo** [comment]
不是很明白你想表达的意思，宽卷积显然针对的是每一个字或词，所以卷积的是embed_size（如300）；fold那里针对每个word-embedding的维度做了个平均，但没什么效果，词向量的单独一个维度应该没什么意义吧（我觉得是这样子）。

**Comment by 473951841** [comment]
是这样的，假设词向量embed_size=300维，首先可以问下最开始处理的embedding output的shape大致是什么样的吗，然后我问的这个fold层问题其实是和之前wide_conv相关的，我在参考您的model相关代码编自己的程序，文章中conv是针对300维的每一维进行操作，我的embedding output输入是（batch_size, sentence_length, 300）,但是经过wide_conv(核心操作是keras的Conv1D)后就变成（batch_size, 宽卷积处理后的sentence_length, filter_num）而300维不见了，所以导致后面fold是依据filter_num来fold的，所以这里想请教一下程序中如何体现出300维的，谢谢您。然后还有一个问题是动态k的选择中句长好像是固定句长最大值，这个是不是每句都不一样？这样在程序中应该如何处理呢，谢谢

**Comment by yongzhuo** [comment]
1. embedding output的shape是(None, 128, 768)   
2. 卷积核是(filter_size(如3,4,5), embed_size(这里就是768))，fold那里其实就是处理filter_num，
    你可以想象一下conv+pool+conv+pool这样的网络，计算机视觉那些出名的网络。
3. 动态池化dynamic_k_max_pooling主要用的是tf.nn.top_k，固定tok就好，与文本长度无关。

具体网络架构你可以看summary输出的结果,如下例子：

Model: "model_1"
____________________________________________________________________________________________________________________________________
Layer (type)                               Output Shape                 Param #         Connected to                                
====================================================================================================================================
input_1 (InputLayer)                       [(None, 128)]                0                                                           
____________________________________________________________________________________________________________________________________
embedding (Embedding)                      (None, 128, 768)             585216          input_1[0][0]                               
____________________________________________________________________________________________________________________________________
wide_convolution_0 (wide_convolution)      (None, 137, 300)             0               embedding[0][0]                             
____________________________________________________________________________________________________________________________________
wide_convolution_1 (wide_convolution)      (None, 133, 300)             0               embedding[0][0]                             
____________________________________________________________________________________________________________________________________
dynamic_k_max_pooling (dynamic_k_max_pooli (None, 85, 300)              0               wide_convolution_0[0][0]                    
____________________________________________________________________________________________________________________________________
dynamic_k_max_pooling_3 (dynamic_k_max_poo (None, 85, 300)              0               wide_convolution_1[0][0]                    
____________________________________________________________________________________________________________________________________
wide_convolution_0_0 (wide_convolution)    (None, 91, 300)              0               dynamic_k_max_pooling[0][0]                 
____________________________________________________________________________________________________________________________________
wide_convolution_1_1 (wide_convolution)    (None, 88, 300)              0               dynamic_k_max_pooling_3[0][0]               
____________________________________________________________________________________________________________________________________
dynamic_k_max_pooling_1 (dynamic_k_max_poo (None, 42, 300)              0               wide_convolution_0_0[0][0]                  
____________________________________________________________________________________________________________________________________
dynamic_k_max_pooling_4 (dynamic_k_max_poo (None, 42, 300)              0               wide_convolution_1_1[0][0]                  
____________________________________________________________________________________________________________________________________
wide_convolution_0_0_0 (wide_convolution)  (None, 46, 300)              0               dynamic_k_max_pooling_1[0][0]               
____________________________________________________________________________________________________________________________________
wide_convolution_1_1_1 (wide_convolution)  (None, 44, 300)              0               dynamic_k_max_pooling_4[0][0]               
____________________________________________________________________________________________________________________________________
prem_fold (prem_fold)                      (None, 46, 150)              0               wide_convolution_0_0_0[0][0]                
____________________________________________________________________________________________________________________________________
prem_fold_1 (prem_fold)                    (None, 44, 150)              0               wide_convolution_1_1_1[0][0]                
____________________________________________________________________________________________________________________________________
dynamic_k_max_pooling_2 (dynamic_k_max_poo (None, 3, 150)               0               prem_fold[0][0]                             
____________________________________________________________________________________________________________________________________
dynamic_k_max_pooling_5 (dynamic_k_max_poo (None, 3, 150)               0               prem_fold_1[0][0]                           
____________________________________________________________________________________________________________________________________
concatenate (Concatenate)                  (None, 6, 150)               0               dynamic_k_max_pooling_2[0][0]               
                                                                                        dynamic_k_max_pooling_5[0][0]               
____________________________________________________________________________________________________________________________________
dropout (Dropout)                          (None, 6, 150)               0               concatenate[0][0]                           
____________________________________________________________________________________________________________________________________
flatten (Flatten)                          (None, 900)                  0               dropout[0][0]                               
____________________________________________________________________________________________________________________________________
dense (Dense)                              (None, 17)                   15317           flatten[0][0]                               
====================================================================================================================================
Total params: 600,533
Trainable params: 600,533
Non-trainable params: 0

**Comment by 473951841** [comment]
您好，我看了下那篇文章，认为folding还是对维度进行的操作，在原始论文的图三中看到folding输入前本来是4维，变成了2维，所以这里在模型中如何体现还不太清楚，然后论文中，决定dynamickmaxpooling 的k 包括一个公式max(s(L-l)/L,ktop),其中s便是输入句长，在您的模型中看得出应该是 select_k函数决定k值，然而s已经固定为len_max，定义在hyper_parameter中，这里是否和原文想表述的不一呢，我的感觉是输入的每个样本句长在进行kmaxpooling时不能先pad成同一长度，这个如何实现还不太清楚。（可以的话希望能加QQ细聊，我的QQ是我的昵称，非常感谢）

**Comment by yongzhuo** [comment]
上图中(batch, seq_len, embed_size)卷积后尺寸变成了(batch, seq_len[wide], filters_num)，这里的filters_num已经不是你认为的词向量维度了。
你可以这样，二维卷积，卷积核为(filter_size, 1)而不是(filter_size, embed_size)，这样子得到的embed_size就会不变，结果会增加一维的filters_num
